project/apps/matching-service/src/db/match.redis.ts

```diff
@@ -47,6 +49,181 @@ export class MatchRedis {
+    }
+  }
+
+  /**
+   * Finds a potential match in redis based on one of the selected categories and complexity
+   * 1. Fetches all the matching matchIds from the sorted set for each category
+   * 2. Checks if the matchId is in the cancelled list
+   * 3. Sort the list of potential matches by earliest timestamp and choose the oldest request
+   * 4. Removes the match request from redis
+   * @param criteria Match requester's selected categories and complexity
+   * @returns The userId and matchId of the potential match if found, otherwise null
+   */
+
+  async findPotentialMatch(criteria: CriteriaDto): Promise<{ userId: string; matchId: string } | null> {
+    const { category, complexity } = criteria;
+
+    const pipeline = this.redisClient.pipeline()
+
+    category.forEach(cat => {
+      const sortedSetKey = `${MATCH_CATEGORY}-${cat}-COMPLEXITY-${complexity}`;
+      pipeline.zrange(sortedSetKey, 0, 0);
+    });
+
+    const responses = await pipeline.exec();
+
+    if (!responses) return null
+
+    const fetchPromises = responses.map(async ([err, res], index) => {
+      if (err) {
+        this.logger.error(`Error fetching matchId(s) from category ${category[index]}:`, err);
+        return null;
+      }
+      const matchId: string = Array.isArray(res) ? res[0] : null;
+      if (!matchId) return null;
+  
+      const matchRequest = await this.getMatchRequest(matchId);
+
+      // Also need to check if the matchId is in the cancelled list
+      const cancelledKey = `${MATCH_CANCELLED_KEY}-${matchId}`;
+      const isCancelled = await this.redisClient.zscore(cancelledKey, matchId);
+
+      if (matchRequest && !isCancelled) {
+        return { matchId, matchRequest };
+      }
+      return null;
+    });
+  
+    const matchResults = await Promise.all(fetchPromises);
+    
+     const validResults = matchResults
+    .filter(req => req !== null);
+
+    if (validResults.length === 0) return null; // No matches found
+
+    // sort by earliest first
+    validResults.sort((a, b) => a.matchRequest.timestamp - b.matchRequest.timestamp);
+    const oldestMatch = validResults[0];
+
+    await this.removeMatchRequest(oldestMatch.matchId);
+
+    return { userId: oldestMatch.matchRequest.userId, matchId: oldestMatch.matchId };
+  }
+
+  /**
+   * Adds a matchId to the cancelled list in redis
+   * @param matchId The matchId to add to the cancelled list
+   */
+
+  async addToCancelledMatchList(matchId: string) {
+    try {
+      const key = `${MATCH_CANCELLED_KEY}-${matchId}`;
+      const timestamp = Date.now();
+
+      await this.redisClient.zadd(key, timestamp, matchId);
+
+      // we can be certain that a matchId would have either been matched or expired within 1 hour
+      await this.redisClient.expire(key, MATCH_EXPIRY_TTL); // 1 hour
+
+      this.logger.log(`Match ${matchId} added to cancelled list`);
+    } catch (error) {
+      this.logger.error(`Error adding match to cancelled list: ${error}`);
+    }
```

Comment:
for this can i ask why we are doing a zadd here, given that a `${MATCH_CANCELLED_KEY}-${matchId}` is unique, would a `.set` be sufficient? and then we can retrieve if the user is cancelled using a `.get`

Reply:
@jeffrey_jian for this one was there some reason behind it? If not, I can change it using the set via cache manager ioredis